### PR TITLE
SI-6928, VerifyError with self reference to super.

### PR DESCRIPTION
**Pull request by paulp**
A bug in typers mishandled varargs. We should get more
aggressive about eliminating all the ad hoc parameter/argument
handling code spread everywhere. For varargs especially:
any code which tries to make an adjustment based on a
repeated parameter is more likely to be wrong than right.
In aggregate these reinventions are a huge source of bugs.

### CODE REVIEW

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 1]
```diff
@@ -2070,25 +2070,29 @@ trait Typers extends Modes with Adaptations with Tags {
...
-          assert(sameLength(args2, params) || call.isErrorTyped, "mismatch " + clazz + " " + (params map (_.tpe)) + " " + args2)//debug
-          (superConstr, args1 ::: args2)
-        case Block(stats, expr) if !stats.isEmpty =>
+          // appending a dummy tree to represent Nil for an empty varargs (is this really necessary?)
+          val applyArgs = if (args.length < params.length) args :+ EmptyTree else args take params.length
+
+          assert(sameLength(applyArgs, params) || call.isErrorTyped, s"arity mismatch but call is not error typed: $clazz (params=$params, args=$applyArgs)")
+          (superConstr, preArgs ::: applyArgs)
+        case Block(stats, expr) if stats.nonEmpty =>
           decompose(stats.last)
```
Comment (xeno-by) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: How about:

```
case Block(_ :+ superCall, _) =>
  decompose(superCall)
```

**Comment by retronym** [comment]
lgtm

**Comment by paulp** [comment]
It came to me in a dream - literally - that this was insufficient. Reviewers, you are the TSA guys who let loaded handguns through security!

**Comment by paulp** [comment]
PLS REBUILD ALL

**Pull request merged**